lib/Conversion/Affine/AffineFullUnroll.cpp

```diff
@@ -0,0 +1,29 @@
+#include "lib/Conversion/Affine/AffineFullUnroll.h"
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Affine/LoopUtils.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "mlir/include/mlir/Pass/Pass.h"
+
+namespace mlir {
+namespace tutorial {
+
+using mlir::affine::AffineForOp;
+using mlir::affine::loopUnrollFull;
+
+struct UnrollAffineLoopPattern : public OpRewritePattern<AffineForOp> {
+  explicit UnrollAffineLoopPattern(MLIRContext *context)
+      : OpRewritePattern(context, 1) {}
+
+  LogicalResult matchAndRewrite(AffineForOp op, PatternRewriter &rewriter) const override {
+    return loopUnrollFull(op);
+  }
+};
+
+void AffineFullUnrollPass::runOnOperation() {
+  RewritePatternSet patterns(&getContext());
```

Comment:
So in this case, if you are already using a rewrite, why not use a dialect conversion driver? (e.g. the `applyPartialConversion`). I think in this case you can use this as a case of "Let's write a pass that removes loop unrolls, thereby removing all affine.for loops". I think that's easier than greedy pattern rewrites, which should be somewhat iterative, and in this case I don't think this is, since the implementation is probably not repeatedly unrolling an inner for loop, but rather loop unrolls a target for loop.

Now I'm wondering how the pattern application for this works since the pattern does converge: does the worklist traverse depth first or breath first on the for loops in the module? (not really tutorial-relevant at this point)

Reply:
The main reason is that I didn't want to introduce too many concepts at once, and I think that dialect conversion comes with a LOT more conceptual overhead than the pattern rewrite. But now that I think about it, I'm more inclined to try just modifying the AST in place with a walk instead of a pattern rewrite.